core/src/test/java/org/elasticsearch/ingest/IngestMetadataTests.java

```diff
@@ -61,4 +66,36 @@ public void testFromXContent() throws IOException {
         assertEquals(pipeline.getConfigAsMap(), m.getPipelines().get("1").getConfigAsMap());
         assertEquals(pipeline2.getConfigAsMap(), m.getPipelines().get("2").getConfigAsMap());
     }
+
+    public void testDiff() throws Exception {
+        Map<String, PipelineConfiguration> pipelines = new HashMap<>();
+        pipelines.put("1", new PipelineConfiguration("1", new BytesArray("{}")));
+        pipelines.put("2", new PipelineConfiguration("2", new BytesArray("{}")));
+        IngestMetadata ingestMetadata1 = new IngestMetadata(pipelines);
+
+        pipelines = new HashMap<>();
+        pipelines.put("1", new PipelineConfiguration("1", new BytesArray("{}")));
+        pipelines.put("3", new PipelineConfiguration("3", new BytesArray("{}")));
+        pipelines.put("4", new PipelineConfiguration("4", new BytesArray("{}")));
+        IngestMetadata ingestMetadata2 = new IngestMetadata(pipelines);
+
+        Diff<MetaData.Custom> diff = ingestMetadata2.diff(ingestMetadata1);
+        IngestMetadata result = (IngestMetadata) diff.apply(ingestMetadata2);
+        assertThat(result, not(equalTo(ingestMetadata1)));
+        assertThat(result.getPipelines().size(), equalTo(3));
+        assertThat(result.getPipelines().containsKey("1"), is(true));
+        assertThat(result.getPipelines().containsKey("3"), is(true));
+        assertThat(result.getPipelines().containsKey("4"), is(true));
+
+        pipelines = new HashMap<>();
+        pipelines.put("1", new PipelineConfiguration("1", new BytesArray("{}")));
+        pipelines.put("2", new PipelineConfiguration("2", new BytesArray("{}")));
+        IngestMetadata ingestMetadata3 = new IngestMetadata(pipelines);
+        diff = ingestMetadata3.diff(ingestMetadata1);
+        result = (IngestMetadata) diff.apply(ingestMetadata3);
+        assertThat(result, equalTo(ingestMetadata1));
+        assertThat(result.getPipelines().size(), equalTo(2));
+        assertThat(result.getPipelines().containsKey("1"), is(true));
+        assertThat(result.getPipelines().containsKey("2"), is(true));
```

Comment:
does it make sense to have a test where the content changes, and not only the id of the pipeline?